### PR TITLE
TIP-110: ANGLE & agEUR redeem

### PR DESCRIPTION
**Pull request by eswak**
- [x] All Tests Passing
- [x] Proposal Added to ProposalsConfig
- [x] Fork Block Correct
- [x] Remove Any .only's on Tests
- [x] Update Documentation If Needed
- [x] Update Roles Config
- [ ] Proposal Submitted

### CODE REVIEW

File: proposals/dao/tip_110.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  pcvStatsBefore = await contracts.collateralizationOracle.pcvStats();
+
+  // angle multisig action : make enough USDC collateral available for redemptions
+  const angleMultisigSigner = await getImpersonatedSigner('0x0C2553e4B9dFA9f83b1A6D3EAB96c4bAaB42d430');
+  await forceEth(angleMultisigSigner.address);
+  await contracts.anglePoolManagerUsdc.connect(angleMultisigSigner).updateStrategyDebtRatio(
+    addresses.angleStrategyUsdc1, // USDC strategy has 57M deployed
+    '0'
+  );
+  await contracts.angleStrategyUsdc1.harvest();
```
Comment (eswak) [inline proposals/dao/tip_110.ts]: This multisig action has to be done by the Angle multisig, coordinate with them so that at least 10.7M USDC is available for redemptions at time of execution

File: proposals/dao/tip_110.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  );
+  // check deposit is empty
+  expect(await contracts.agEurUniswapPCVDeposit.balance()).to.be.equal('0');
+  expect(await contracts.fei.balanceOf(addresses.agEurUniswapPCVDeposit)).to.be.equal('0');
+  expect(await contracts.agEUR.balanceOf(addresses.agEurUniswapPCVDeposit)).to.be.equal('0');
+
+  // check redemptions of agEUR > DAI
+  const daiRedeemed = (await contracts.dai.balanceOf(addresses.daiFixedPricePSM)).sub(daiBalanceBefore);
+  console.log('daiRedeemed', daiRedeemed.toString() / 1e18);
+  expect(daiRedeemed).to.be.at.least(ethers.utils.parseEther('9500000')); // >9.5M DAI
```
Comment (eswak) [inline proposals/dao/tip_110.ts]: the expected behavior, if enough USDC is available for redemptions, is to convert all agEUR to DAI with a fixed fee of 0.45% (Angle Protocol redemption) + 0% (Maker USDC->DAI PSM)

File: contracts/pcv/angle/AngleEuroRedeemer.sol [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            agEurSpentToRedeemUsdc,
+            address(this),
+            address(this),
+            ANGLE_POOLMANAGER_USDC,
+            minUsdcOut
+        );
+
+        // Use Maker PSM to convert USDC to DAI
+        address daiUsdcPSMAuth = 0x0A59649758aa4d66E25f08Dd01271e891fe52199;
+        address daiUsdcPSM = 0x89B78CfA322F6C5dE0aBcEecab66Aee45393cC5A;
```
Comment (Joeysantoro) [inline contracts/pcv/angle/AngleEuroRedeemer.sol]: these should be constants at top of file

File: contracts/pcv/angle/AngleEuroRedeemer.sol [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    // Tribe DAO addresses
+    IOracle public constant TRIBEDAO_EUR_USD_ORACLE = IOracle(0xFb3a062236A7E08b572F17bc9Ad2bBc2becB87b1);
+    address public constant TRIBEDAO_FEI_DAI_PSM = 0x2A188F9EB761F70ECEa083bA6c2A40145078dfc2;
+    address public constant TRIBEDAO_TC_TIMELOCK = 0xe0C7DE94395B629860Cbb3c42995F300F56e6d7a;
+
+    /// @notice redeem all agEUR held on this contract to USDC using Angle Protocol,
+    /// and then use the Maker PSM to convert the USDC to DAI, and send the DAI to
+    /// Tribe DAO's FEI/DAI PSM.
+    function redeem() external {
```
Comment (Joeysantoro) [inline contracts/pcv/angle/AngleEuroRedeemer.sol]: should this be access controlled?

Comment (eswak) [reply contracts/pcv/angle/AngleEuroRedeemer.sol]: I can put onlyGuardianOrGovernor but the contract is meant to be temporary (it's going to be useless after dao vote execution, so it is funded & atomically redeem() is called)

Comment (Joeysantoro) [reply contracts/pcv/angle/AngleEuroRedeemer.sol]: better to not leave anything to chance, access control if its fine to do so

Comment (Joeysantoro) [reply contracts/pcv/angle/AngleEuroRedeemer.sol]: same with the withdraw function

File: contracts/pcv/angle/AngleEuroRedeemer.sol [hunk 5]
```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import "./IStableMaster.sol";
+import "../../external/Decimal.sol";
```
Comment (thomas-waite) [inline contracts/pcv/angle/AngleEuroRedeemer.sol]: Style preference, but I much prefer imports to be explicit, `import {X} from "../../Y.sol"` - I've had cases before where it's been non-obvious what is being imported into scope and had to dig down to see where a contract comes from

File: contracts/pcv/angle/AngleEuroRedeemer.sol [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+    // Tribe DAO addresses
+    IOracle public constant TRIBEDAO_EUR_USD_ORACLE = IOracle(0xFb3a062236A7E08b572F17bc9Ad2bBc2becB87b1);
+    address public constant TRIBEDAO_FEI_DAI_PSM = 0x2A188F9EB761F70ECEa083bA6c2A40145078dfc2;
+    address public constant TRIBEDAO_TC_TIMELOCK = 0xe0C7DE94395B629860Cbb3c42995F300F56e6d7a;
+
+    /// @notice redeem all agEUR held on this contract to USDC using Angle Protocol,
+    /// and then use the Maker PSM to convert the USDC to DAI, and send the DAI to
+    /// Tribe DAO's FEI/DAI PSM.
+    function redeem() external onlyGuardianOrGovernor {
```
Comment (thomas-waite) [inline contracts/pcv/angle/AngleEuroRedeemer.sol]: Would be nice if the TribalCouncil was able to call this, but we don't really have an appropriate role setup for selling/converting PCV. 

Closest would probably be `PCV_SAFE_MOVER_ROLE`

File: proposals/dao/tip_118.ts [hunk 7]
```diff
@@ -68,6 +78,26 @@ const deploy: DeployUpgradeFunc = async (deployAddress: string, addresses: Named
...
+  // make sure ETH oracle is fresh (for B.AMM not to revert, etc)
+  // Read Chainlink ETHUSD price & override chainlink storage to make it a fresh value
+  const ethPrice = (await contracts.chainlinkEthUsdOracleWrapper.read())[0].toString() / 1e10;
+  await overwriteChainlinkAggregator(addresses.chainlinkEthUsdOracle, Math.round(ethPrice).toString(), '8');
+
+  // read pcvStats before proposal execution
+  pcvStatsBefore = await contracts.collateralizationOracle.pcvStats();
+
+  // angle multisig action : make enough USDC collateral available for redemptions
+  const angleMultisigSigner = await getImpersonatedSigner('0x0C2553e4B9dFA9f83b1A6D3EAB96c4bAaB42d430');
```
Comment (thomas-waite) [inline proposals/dao/tip_118.ts]: Would probably add this address to `mainnetAddresses.ts`

File: proposals/dao/tip_118.ts [hunk 8]
```diff
@@ -68,6 +78,26 @@ const deploy: DeployUpgradeFunc = async (deployAddress: string, addresses: Named
...
+  await forceEth(angleMultisigSigner.address);
```
Comment (thomas-waite) [inline proposals/dao/tip_118.ts:103]: Is there guaranteed to be enough USDC collateral available for redemptions on-chain when this executes?

File: contracts/pcv/angle/AngleEuroRedeemer.sol [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+        // redeem USDC available
+        uint256 usdcAvailableForRedeem = ANGLE_POOLMANAGER_USDC.getBalance();
+        // scale decimals 6 -> 18
+        uint256 agEurSpentToRedeemUsdc = (usdcAvailableForRedeem * 1e12 * 1e18) / (usdPerEur);
+        if (agEurSpentToRedeemUsdc > agEurBalance) {
+            agEurSpentToRedeemUsdc = agEurBalance;
+        }
+        // no need to check stableMaster.collateralMap[PoolManager].stocksUsers because
+        // USDC has a lot of stock available (~57M)
```
Comment (thomas-waite) [inline contracts/pcv/angle/AngleEuroRedeemer.sol:64]: In the case where there isn't enough USDC, `redeem()` will just revert right, so no problem if the USDC is pulled

File: contracts/pcv/angle/AngleEuroRedeemer.sol [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+import "../../refs/CoreRef.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+interface IMakerPSM {
+    function sellGem(address, uint256) external;
+}
+
+/// @title Contract to redeem agEUR for FEI and DAI
+/// @author eswak
+contract AngleEuroRedeemer is CoreRef {
```
Comment (thomas-waite) [inline contracts/pcv/angle/AngleEuroRedeemer.sol:17]: This is a fair amount of Solidity to be shipping without tests/integration tests. 

Is it practical to write an integration test that validates we receive the expected amount of DAI out for an amount of Angle in?

**Review by thomas-waite — COMMENTED** [review]
Really like the idea of this contract and it looks good. 

If it's practical and there's time, would like to see an integration test that shows we receive the expected amount of DAI for an input amount of ANGLE; but no strong feelings. Even if something goes wrong we could pull the funds from the contract using `withdrawERC20()`

File: contracts/pcv/angle/AngleEuroRedeemer.sol (hunk 5 above)
Comment (eswak) [reply contracts/pcv/angle/AngleEuroRedeemer.sol]: I'll update the imports

File: contracts/pcv/angle/AngleEuroRedeemer.sol (hunk 6 above)
Comment (eswak) [reply contracts/pcv/angle/AngleEuroRedeemer.sol]: I'll gate the call to
```
hasAnyOfThreeRoles(
    TribeRoles.GOVERNOR,
    TribeRoles.GUARDIAN,
    TribeRoles.PCV_SAFE_MOVER_ROLE
)
```

File: proposals/dao/tip_118.ts (hunk 7 above)
Comment (eswak) [reply proposals/dao/tip_118.ts]: adding to mainnetAddresses

File: proposals/dao/tip_118.ts (hunk 8 above)
Comment (eswak) [reply proposals/dao/tip_118.ts:103]: not really, right now there is 57M USDC available but the Angle team have to make it available for withdrawal (we shall coordinate to do this before the dao vote executes)

If there is not enough available for redeem (let's say, 1M USDC is available), the agEUR redeemer will burn ~950k agEUR to redeem the 1M USDC available and will continue to hold ~8.5M agEUR waiting to be redeemed. This is not a problem and we can redeem in multiple steps but the agEUR won't show in the CR oracle as long as it's not redeemed

File: contracts/pcv/angle/AngleEuroRedeemer.sol (hunk 9 above)
Comment (eswak) [reply contracts/pcv/angle/AngleEuroRedeemer.sol:64]: yep it will revert and we'll have to find another way to redeem this agEUR :-/

File: contracts/pcv/angle/AngleEuroRedeemer.sol (hunk 10 above)
Comment (eswak) [reply contracts/pcv/angle/AngleEuroRedeemer.sol:17]: The proposal script itself validates at the end that 9,556,922 DAI is redeemed but I'll write an integration test for various amounts & the special cases where not enough is available for redeem etc

File: proposals/description/tip_118.ts [hunk 11]
```diff
@@ -308,16 +308,105 @@ const tip_118: TemplatedProposalDescription = {
...
+    },
+    {
+      target: 'collateralizationOracle',
+      values: '0',
+      method: 'removeDeposits(address[])',
+      arguments: (addresses) => [[addresses.agEurUniswapPCVDeposit, addresses.uniswapLensAgEurUniswapGauge]],
+      description: 'Remove agEUR addresses from CR oracle'
+    },
+    {
+      target: 'pcvGuardian',
```
Comment (thomas-waite) [inline proposals/description/tip_118.ts]: This should be `pcvGuardianNew`. I introduced this anti-pattern, should update to use a versioning naming scheme like `pcvGuardianV2`, `pcvGuardianV1` etc. but that can be part of a different PR

Comment (eswak) [reply proposals/description/tip_118.ts]: wow good catch
